src/coreclr/vm/eventtrace.cpp

```diff
@@ -2328,6 +2328,72 @@ enum CallbackProviderIndex
     DotNETRuntimePrivate = 3
 };
 
+// EventFilterType identifies the filter type used by the PEVENT_FILTER_DESCRIPTOR
+enum EventFilterType
+{
+    None = 0x0,
+    ClientSequenceNumber = 0x1,
+    EventPipePayload = 0x2,
+    // Schematized = 0x80000000,
+    // SystemFlags = 0x80000001,
+    // TraceHandle = 0x80000002,
+    // PID = 0x80000004,
+    // ExecutableName = 0x80000008,
+    // PackageID = 0x80000010,
+    // PackageAppID = 0x80000020,
+    // Payload = 0x80000100,
+    // EventID = 0x80000200,
+    // EventName = 0x80000400,
+    // Stackwalk = 0x80001000,
+    // StackwalkName = 0x80002000,
+    // StackwalkLevelKW = 0x80004000,
+};
+
+VOID ParseFilterDataClientSequenceNumber(
+    PEVENT_FILTER_DESCRIPTOR FilterData,
+    LONGLONG * pClientSequenceNumber)
+{
+    LIMITED_METHOD_CONTRACT;
+
+#if !defined(HOST_UNIX)
+    if (FilterData == NULL)
+        return;
+
+    if (FilterData->Type == ClientSequenceNumber && FilterData->Size == sizeof(LONGLONG))
+    {
+        *pClientSequenceNumber = *(LONGLONG *) (FilterData->Ptr);
+    }
+    else if (FilterData->Type == EventPipePayload)
+    {
+        const char* buffer = reinterpret_cast<const char*>(FilterData->Ptr);
+        const char* buffer_end = buffer + FilterData->Size;
+
+        while (buffer < buffer_end)
+        {
+            const char* key = buffer;
+            buffer += strlen(key) + 1;
+
+            if (buffer >= buffer_end)
+                break;
+
+            const char* value = buffer;
+            buffer += strlen(value) + 1;
+
```

Comment:
To be safe against a mal-formed buffer we should do a bounds check on the value portion too:
```suggestion
if(buffer > buffer_end)
    break;
```